File: Algo-1/week2/3-Quadruplets/quadruplets.cpp
#include <iostream>
#include <algorithm>
#include <vector>


class Quadruplets {
public:
	// Returns the number of quadruplets that sum to zero
	int zeroQuadrupletsCount(std::vector<int>& a, std::vector<int>& b, std::vector<int>& c, std::vector<int>& d) {
		
		int counter = 0;

		std::vector<int> AandBpermutation;

		for (int i = 0; i < a.size(); i++) {
			for (int j = 0; j < b.size(); j++) {
				AandBpermutation.push_back(a[i] + b[j]);
			}
		}

		std::sort(AandBpermutation.begin(), AandBpermutation.end());

		for (int i = 0; i < c.size(); i++) {
			for (int j = 0; j < d.size(); j++) {
				int sum = -(c[i] + d[j]);
				counter += std::upper_bound(AandBpermutation.begin(), AandBpermutation.end(), sum) - std::lower_bound(AandBpermutation.begin(), AandBpermutation.end(), sum);
			}
		}

		return counter;
	}
};

int main() {

	int n, temp;
	std::cin >> n;
	std::vector<int> a, b, c, d;
	
	Quadruplets test;

	for (int i = 0; i < n; i++) {
		std::cin >> temp;
		a.push_back(temp);
	}

	for (int i = 0; i < n; i++) {
		std::cin >> temp;
		b.push_back(temp);
	}

	for (int i = 0; i < n; i++) {
		std::cin >> temp;
		c.push_back(temp);
	}

	for (int i = 0; i < n; i++) {
		std::cin >> temp;
		d.push_back(temp);
	}
	std::cout << test.zeroQuadrupletsCount(a, b, c, d) << std::endl;
	
	return 0;
}

File: Algo-1/week2/3-Quadruplets/quadruplets.py
# Returns the number of quadruplets that sum to zero.
# a - [int]
# b - [int]
# d - [int]
# c - [int]


# a + b == -(c + d)
# sort
# binary search


def zero_quadruplets_count(a, b, c, d):
    left_sums = []
    right_sums = []
    result = 0

    for i in a:
        for j in b:
            left_sums.append(i + j)

    for i in c:
        for j in d:
            right_sums.append(i + j)

    for i in left_sums:
        for j in right_sums:
            if i + j == 0:
                result += 1

    return result


def main():

    a = [5, 3, 4]
    b = [-2, -1, 6]
    c = [-1, -2, 4]
    d = [-1, -2, 7]

    print(zero_quadruplets_count(a, b, c, d))


if __name__ == '__main__':
    main()
